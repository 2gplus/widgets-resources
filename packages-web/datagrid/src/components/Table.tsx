import { createElement, Dispatch, ReactElement, ReactNode, SetStateAction, useMemo } from "react";
import { ObjectItem } from "mendix";
import { ColumnSelector } from "./ColumnSelector";
import { ClientSidePagination, Pagination } from "./Pagination";
import { Header } from "./Header";
import { InfiniteBody } from "./InfiniteBody";
import {
    ColumnInterface,
    ColumnWithStrictAccessor,
    FilterValue,
    IdType,
    Row,
    useColumnOrder,
    useFilters,
    useFlexLayout,
    usePagination,
    useResizeColumns,
    useSortBy,
    useTable
} from "react-table";
import { ColumnsPreviewType, ColumnsType } from "../../typings/DatagridProps";

export interface TableProps {
    className: string;
    data: ObjectItem[];
    columns: Array<ColumnsType | ColumnsPreviewType>;
    headerWidgets?: ReactNode;
    footerWidgets?: ReactNode;
    columnsFilterable: boolean;
    columnsSortable: boolean;
    columnsResizable: boolean;
    columnsDraggable: boolean;
    columnsHidable: boolean;
    paging: boolean;
    page: number;
    pageSize: number;
    pagingPosition: string;
    setPage?: Dispatch<SetStateAction<number>>;
    hasMoreItems: boolean;
}

interface TableRowData {
    item: ObjectItem;
}

export function Table({
    // todo: remove decomposition
    className,
    columns,
    data,
    headerWidgets,
    footerWidgets,
    columnsFilterable,
    columnsSortable,
    columnsResizable,
    columnsDraggable,
    columnsHidable,
    paging,
    pageSize,
    pagingPosition,
    hasMoreItems,
    page,
    setPage
}: TableProps): ReactElement {
    const isSortingOrFiltering = columnsFilterable || columnsSortable;
    const isInfinite = !paging && !isSortingOrFiltering;

    const filterTypes = {
        text: (rows: Array<Row<object>>, id: IdType<object>, filterValue: FilterValue) => {
            return rows.filter(row => {
                const rowValue = row.values[id];
                return rowValue !== undefined
                    ? String(rowValue)
                          .toLowerCase()
                          .startsWith(String(filterValue).toLowerCase())
                    : true;
            });
        }
    };
    const tableColumns: Array<ColumnWithStrictAccessor<TableRowData>> = useMemo(
        () =>
            columns.map((column, index) => ({
                id: index.toString(),
                accessor: "item",
                Header: typeof column.header === "object" ? column.header.value : column.header,
                filter: "text",
                isVisible: column.hidable !== "hidden",
                canHide: column.hidable !== "no",
                canDrag: column.draggable,
                disableSortBy: !column.sortable,
                disableResizing: !column.resizable,
                disableFilters: !column.filterable,
                Cell: ({ value }) => {
                    if (column.hasWidgets && column.content) {
                        return "renderer" in column.content ? (
                            <column.content.renderer>
                                <div />
                                {/* TODO: Not just a div */}
                            </column.content.renderer>
                        ) : (
                            <div>{column.content(value)}</div>
                        );
                    }
                    return column.attribute instanceof Function
                        ? column.attribute(value).displayValue
                        : column.attribute;
                }
            })),
        [columns]
    );

    const defaultColumn: ColumnInterface<TableRowData> = useMemo(
        () => ({
            Filter: ({ column: { filterValue, setFilter, filter } }): ReactElement => (
                <input
                    className="form-control"
                    value={filterValue || ""}
                    onChange={e => {
                        setFilter(e.target.value || undefined); // Set undefined to remove the filter entirely
                    }}
                    placeholder={`Search ${filter ? filter : ""}...`}
                />
            ),
            ...(columnsResizable
                ? {
                      width: 150, // width is used for both the flex-basis and flex-grow
                      maxWidth: 200, // maxWidth is only used as a limit for resizing
                      minWidth: 15 // minWidth is only used as a limit for resizing
                  }
                : {})
        }),
        [columns]
    );

    const {
        getTableProps,
        headerGroups,
        rows,
        page: rowsPagination,
        prepareRow,
        getTableBodyProps,
        allColumns,
        setColumnOrder,
        visibleColumns,
        pageOptions,
        state: { pageIndex },
        gotoPage,
        previousPage,
        nextPage,
        canPreviousPage,
        canNextPage
    } = useTable<TableRowData>(
        {
            columns: tableColumns,
            data: useMemo(() => data.map(item => ({ item })), [data]),
            defaultColumn,
            filterTypes,
            disableResizing: !columnsResizable,
            disableSortBy: !columnsSortable,
            disableFilters: !columnsFilterable,
            initialState: { pageSize },
            disableMultiSort: true
        },
        useFilters,
        useSortBy,
        usePagination, // Used for client side pagination
        useColumnOrder,
        useFlexLayout,
        useResizeColumns
    );

    return (
        <div className={className}>
            <div className="thead">
                {headerWidgets}
                {columnsHidable && <ColumnSelector allColumns={allColumns} />}
            </div>
            <div {...getTableProps()} className="table">
                <div role="rowgroup" className="thead">
                    {!isInfinite && !isSortingOrFiltering && paging && pagingPosition === "top" && setPage && (
                        <div className="tr">
                            <div className="td">
                                <Pagination page={page} setPage={setPage} hasMoreItems={hasMoreItems} />
                            </div>
                        </div>
                    )}
                    {isSortingOrFiltering && paging && pagingPosition === "top" && (
                        <div className="tr">
                            <div className="td">
                                <ClientSidePagination
                                    canNextPage={canNextPage}
                                    canPreviousPage={canPreviousPage}
                                    gotoPage={gotoPage}
                                    nextPage={nextPage}
                                    page={pageIndex}
                                    pageOptions={pageOptions}
                                    previousPage={previousPage}
                                />
                            </div>
                        </div>
                    )}
                    {headerGroups.map((headerGroup, index: number) => (
                        <div {...headerGroup.getHeaderGroupProps({})} key={`headers_row_${index}`} className="tr">
                            {headerGroup.headers.map((column, index) => (
                                <Header
                                    column={column}
                                    key={`headers_column_${index}`}
                                    sortable={columnsSortable}
                                    resizable={columnsResizable}
                                    filterable={columnsFilterable}
                                    draggable={columnsDraggable}
                                    visibleColumns={visibleColumns}
                                    setColumnOrder={setColumnOrder}
                                />
                            ))}
                        </div>
                    ))}
                </div>
                <InfiniteBody
                    isInfinite={isInfinite}
                    hasMoreItems={hasMoreItems}
                    setPage={setPage}
                    {...getTableBodyProps()}
                >
                    {(isSortingOrFiltering && paging ? rowsPagination : rows).map((row, index) => {
                        prepareRow(row);
                        return (
                            <div {...row.getRowProps()} key={`row_${index}`} className="tr">
                                {row.cells.map((cell, index) => (
                                    <div
                                        {...cell.getCellProps()}
                                        {...(!columnsResizable ? { style: { flex: "1 1 0px" } } : {})}
                                        key={`column_${index}`}
                                        className="td"
                                    >
                                        {cell.render("Cell")}
                                    </div>
                                ))}
                            </div>
                        );
                    })}
                </InfiniteBody>
                <div className="tfoot">
                    {!isInfinite && !isSortingOrFiltering && paging && pagingPosition === "bottom" && setPage && (
                        <div className="tr">
                            <div className="td">
                                <Pagination page={page} setPage={setPage} hasMoreItems={hasMoreItems} />
                            </div>
                        </div>
                    )}
                    {isSortingOrFiltering && paging && pagingPosition === "bottom" && (
                        <div className="tr">
                            <div className="td">
                                <ClientSidePagination
                                    canNextPage={canNextPage}
                                    canPreviousPage={canPreviousPage}
                                    gotoPage={gotoPage}
                                    nextPage={nextPage}
                                    page={pageIndex}
                                    pageOptions={pageOptions}
                                    previousPage={previousPage}
                                />
                            </div>
                        </div>
                    )}
                    {footerWidgets}
                </div>
            </div>
        </div>
    );
}
