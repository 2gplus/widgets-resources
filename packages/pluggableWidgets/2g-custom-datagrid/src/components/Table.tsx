import { createElement, ReactElement, useState } from "react";
import { ColumnsType, DefaultTriggerEnum, SelectionModeEnum } from "../../typings/DataGridProps";
import { ObjectItem, ListActionValue } from "mendix";
import { getGridTemplateColumns } from "../utils";
import { Header } from "./Header";
import { Row } from "./Row";
import { MouseEvent } from "react";

export interface TableProps {
    data: ObjectItem[];
    className: string;
    columns: ColumnsType[];
    onTrigger?: ListActionValue;
    defaultTrigger: DefaultTriggerEnum;
    selectionMode: SelectionModeEnum;
}

export const Table = (props: TableProps): ReactElement => {
    const isMultiSelect = props.selectionMode === "multi";
    const gridTemplateColumns = getGridTemplateColumns(props.columns, isMultiSelect);
    const [selection, setSelection] = useState<ObjectItem[]>([]);

    const handleCheckboxClick = (e: MouseEvent<HTMLInputElement>, item?: ObjectItem) => {
        const { name, checked } = e.currentTarget;
        console.log(item);
        if (checked) {
            if (name === "allSelect") {
                setSelection(props.data);
            } else if (item !== undefined) {
                setSelection([...selection, item]);
            }
        } else {
            if (name === "allSelect") {
                setSelection([]);
            } else if (item !== undefined) {
                console.log(item.id);
                let s = selection.filter(s => s.id !== item.id);
                setSelection(s);
            }
        }
    };

    const rows: ReactElement[] = props.data.map(row => {
        return (
            <Row
                checkbox={isMultiSelect}
                columns={props.columns}
                handleCheckboxClick={handleCheckboxClick}
                selection={selection}
                defaultTrigger={props.defaultTrigger}
                item={row}
                onTrigger={props.onTrigger}
            />
        );
    });

    return (
        <div className={"dg-grid " + props.className} role="table">
            <div className="dg-content" role="rowgroup" style={{ gridTemplateColumns }}>
                <Header
                    columns={props.columns}
                    checkbox={isMultiSelect}
                    handleCheckboxClick={handleCheckboxClick}
                    selection={selection}
                    data={props.data}
                />
                {rows}
            </div>
        </div>
    );
};
