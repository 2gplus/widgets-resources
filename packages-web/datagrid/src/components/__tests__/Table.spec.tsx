import { shallow } from "enzyme";
import { createElement } from "react";
import { Table, TableProps } from "../Table";
import { mockColumns, mockColumnsConfig, mockData } from "./mock";

describe("Table", () => {
    it("renders the structure correctly", () => {
        const component = shallow(<Table {...mockTableProps()} />);

        expect(component).toMatchSnapshot();
    });

    it("renders the structure correctly with sorting", () => {
        const component = shallow(<Table {...mockTableProps()} columnsSortable />);

        expect(component).toMatchSnapshot();
    });

    it("renders the structure correctly with resizing", () => {
        const component = shallow(<Table {...mockTableProps()} columnsResizable />);

        expect(component).toMatchSnapshot();
    });

    it("renders the structure correctly with dragging", () => {
        const component = shallow(<Table {...mockTableProps()} columnsDraggable />);

        expect(component).toMatchSnapshot();
    });

    it("renders the structure correctly with filtering", () => {
        const component = shallow(<Table {...mockTableProps()} columnsFilterable />);

        expect(component).toMatchSnapshot();
    });

    it("renders the structure correctly with hiding", () => {
        const component = shallow(<Table {...mockTableProps()} columnsHidable />);

        expect(component).toMatchSnapshot();
    });

    it("renders the structure correctly with paging", () => {
        const component = shallow(<Table {...mockTableProps()} paging />);

        expect(component).toMatchSnapshot();
    });
});

function mockTableProps(): TableProps {
    return {
        setPage: jest.fn(),
        page: 1,
        hasMoreItems: false,
        pageSize: 10,
        columnsResizable: false,
        paging: false,
        pagingPosition: "bottom",
        columnsHidable: false,
        columnsDraggable: false,
        footerWidgets: undefined,
        headerWidgets: undefined,
        className: "test",
        columnsFilterable: false,
        columnsSortable: false,
        columnDefinitions: mockColumns(5),
        columnsConfig: mockColumnsConfig(5),
        data: mockData(5)
    };
}
