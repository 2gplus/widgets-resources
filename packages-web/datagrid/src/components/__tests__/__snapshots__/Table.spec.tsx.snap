// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table renders the structure correctly 1`] = `
<div
  className="test"
>
  <div
    className="header"
  />
  <div
    className="table"
    role="table"
    style={
      Object {
        "minWidth": "0px",
      }
    }
  >
    <div
      className="thead"
      role="rowgroup"
    />
    <InfiniteBody
      hasMoreItems={false}
      isInfinite={true}
      role="rowgroup"
      setPage={[MockFunction]}
    />
    <div
      className="tfoot"
    />
  </div>
</div>
`;

exports[`Table renders the structure correctly with dragging 1`] = `
<div
  className="test"
>
  <div
    className="header"
  />
  <div
    className="table"
    role="table"
    style={
      Object {
        "minWidth": "0px",
      }
    }
  >
    <div
      className="thead"
      role="rowgroup"
    />
    <InfiniteBody
      hasMoreItems={false}
      isInfinite={true}
      role="rowgroup"
      setPage={[MockFunction]}
    />
    <div
      className="tfoot"
    />
  </div>
</div>
`;

exports[`Table renders the structure correctly with filtering 1`] = `
<div
  className="test"
>
  <div
    className="header"
  />
  <div
    className="table"
    role="table"
    style={
      Object {
        "minWidth": "0px",
      }
    }
  >
    <div
      className="thead"
      role="rowgroup"
    />
    <InfiniteBody
      hasMoreItems={false}
      isInfinite={false}
      role="rowgroup"
      setPage={[MockFunction]}
    />
    <div
      className="tfoot"
    />
  </div>
</div>
`;

exports[`Table renders the structure correctly with hiding 1`] = `
<div
  className="test"
>
  <div
    className="header"
  >
    <ColumnSelector
      allColumns={Array []}
      columnsConfig={Object {}}
    />
  </div>
  <div
    className="table"
    role="table"
    style={
      Object {
        "minWidth": "0px",
      }
    }
  >
    <div
      className="thead"
      role="rowgroup"
    />
    <InfiniteBody
      hasMoreItems={false}
      isInfinite={true}
      role="rowgroup"
      setPage={[MockFunction]}
    />
    <div
      className="tfoot"
    />
  </div>
</div>
`;

exports[`Table renders the structure correctly with paging 1`] = `
<div
  className="test"
>
  <div
    className="header"
  />
  <div
    className="table"
    role="table"
    style={
      Object {
        "minWidth": "0px",
      }
    }
  >
    <div
      className="thead"
      role="rowgroup"
    />
    <InfiniteBody
      hasMoreItems={false}
      isInfinite={false}
      role="rowgroup"
      setPage={[MockFunction]}
    />
    <div
      className="tfoot"
    >
      <div
        className="tr"
      >
        <div
          className="td"
        >
          <Pagination
            hasMoreItems={false}
            page={1}
            setPage={[MockFunction]}
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Table renders the structure correctly with resizing 1`] = `
<div
  className="test"
>
  <div
    className="header"
  />
  <div
    className="table"
    role="table"
    style={
      Object {
        "minWidth": "0px",
      }
    }
  >
    <div
      className="thead"
      role="rowgroup"
    />
    <InfiniteBody
      hasMoreItems={false}
      isInfinite={true}
      role="rowgroup"
      setPage={[MockFunction]}
    />
    <div
      className="tfoot"
    />
  </div>
</div>
`;

exports[`Table renders the structure correctly with sorting 1`] = `
<div
  className="test"
>
  <div
    className="header"
  />
  <div
    className="table"
    role="table"
    style={
      Object {
        "minWidth": "0px",
      }
    }
  >
    <div
      className="thead"
      role="rowgroup"
    />
    <InfiniteBody
      hasMoreItems={false}
      isInfinite={false}
      role="rowgroup"
      setPage={[MockFunction]}
    />
    <div
      className="tfoot"
    />
  </div>
</div>
`;
